In [18]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

__In order to perform critic's preferences analysis according to film genres we need to match the genres from the file 'movie_lens_movies', because data scrapped from the rottentomatoes webpage does not contain genres:__

In [19]:
genres=pd.read_csv('movie_lens_movies.csv')
genres=genres.reset_index(drop=True)
genres.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [20]:
genres.shape

(27278, 3)

In [21]:
genres['title']=genres.title.str[:-7]
genres.head()

movieId                        title  \
0        1                    Toy Story   
1        2                      Jumanji   
2        3             Grumpier Old Men   
3        4            Waiting to Exhale   
4        5  Father of the Bride Part II   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [22]:
genres.columns=['ID','Title','Genres']
genres.head()

ID                        Title  \
0   1                    Toy Story   
1   2                      Jumanji   
2   3             Grumpier Old Men   
3   4            Waiting to Exhale   
4   5  Father of the Bride Part II   

                                        Genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [23]:
all_critic_opinions=pd.read_csv('all_critic_opinions_file.csv')
all_critic_opinions=all_critic_opinions.drop('Unnamed: 0',axis=1)
all_critic_opinions=all_critic_opinions.reset_index(drop=True)
all_critic_opinions.shape

(670325, 5)

In [24]:
all_critic_opinions.head()

CRating TMeter                          Title  \
0     3/5    87%    Ant-Man and the Wasp (2018)   
1   3.5/5    67%               Ocean's 8 (2018)   
2   4.5/5    83%              Deadpool 2 (2018)   
3     NaN    83%  Avengers: Infinity War (2018)   
4   4.5/5    97%           Black Panther (2018)   

                                              Review            Critic  
0  Ant-Man and the Wasp is an airy, nimble piece ...  alex-abad-santos  
1  Ocean's 8, at its most endearing, is a slick, ...  alex-abad-santos  
2  A delirious exhibition for the bawdy, acerbic ...  alex-abad-santos  
3  For the first time in a while, I can't wait to...  alex-abad-santos  
4  Coogler and his talented cast and team tell a ...  alex-abad-santos

In [25]:
all_critic_opinions=all_critic_opinions.drop(all_critic_opinions[all_critic_opinions['Review']=='Review'].index)
all_critic_opinions=all_critic_opinions.drop(all_critic_opinions[all_critic_opinions['TMeter']=='No Score Yet'].index)

In [26]:
all_critic_opinions['Year']=all_critic_opinions.Title.str[-5:-1]

In [27]:
all_critic_opinions['TMeter']=all_critic_opinions.TMeter.str[:-1]
all_critic_opinions['Title']=all_critic_opinions.Title.str[:-7]

all_critic_opinions['TMeter']=pd.to_numeric(all_critic_opinions['TMeter'])/100
all_critic_opinions.head()

CRating  TMeter                   Title  \
0     3/5    0.87    Ant-Man and the Wasp   
1   3.5/5    0.67               Ocean's 8   
2   4.5/5    0.83              Deadpool 2   
3     NaN    0.83  Avengers: Infinity War   
4   4.5/5    0.97           Black Panther   

                                              Review            Critic  Year  
0  Ant-Man and the Wasp is an airy, nimble piece ...  alex-abad-santos  2018  
1  Ocean's 8, at its most endearing, is a slick, ...  alex-abad-santos  2018  
2  A delirious exhibition for the bawdy, acerbic ...  alex-abad-santos  2018  
3  For the first time in a while, I can't wait to...  alex-abad-santos  2018  
4  Coogler and his talented cast and team tell a ...  alex-abad-santos  2018

In [28]:
all_critic_opinions=all_critic_opinions.drop('Review',axis=1)

In [29]:
all_critic_opinions.head()

CRating  TMeter                   Title            Critic  Year
0     3/5    0.87    Ant-Man and the Wasp  alex-abad-santos  2018
1   3.5/5    0.67               Ocean's 8  alex-abad-santos  2018
2   4.5/5    0.83              Deadpool 2  alex-abad-santos  2018
3     NaN    0.83  Avengers: Infinity War  alex-abad-santos  2018
4   4.5/5    0.97           Black Panther  alex-abad-santos  2018

In [30]:
all_critic_opinions.shape

(532012, 5)

In [31]:
len(all_critic_opinions['Title'].unique())

21981

__Merging the two files on Title:__

In [32]:
all_critic_opinions1=pd.merge(all_critic_opinions,genres,on='Title')
all_critic_opinions1.head()

CRating  TMeter         Title                Critic  Year      ID  \
0   3.5/5    0.92  Wonder Woman      alex-abad-santos  2017  108795   
1     NaN    0.92  Wonder Woman        beth-accomando  2017  108795   
2     3/4    0.92  Wonder Woman          thelma-adams  2017  108795   
3     4/5    0.92  Wonder Woman  sergio-lopez-aguirre  2017  108795   
4     NaN    0.92  Wonder Woman    victoria-alexander  2017  108795   

                               Genres  
0  Action|Adventure|Animation|Fantasy  
1  Action|Adventure|Animation|Fantasy  
2  Action|Adventure|Animation|Fantasy  
3  Action|Adventure|Animation|Fantasy  
4  Action|Adventure|Animation|Fantasy

In [33]:
all_critic_opinions1.shape

(271354, 7)

In [34]:
len(all_critic_opinions1['Title'].unique())

8681

In [35]:
len(all_critic_opinions1['Critic'].unique())

1822

In [36]:
len(all_critic_opinions1['Genres'].unique())

833

__The new file contains 833 film genres which have to be grouped to a much lower number of genres:__

In [57]:
all_critic_opinions1['Genres'].value_counts()[:50]

Drama                               37028
Comedy                              17110
Documentary                         14117
Comedy|Drama                        14077
Drama|Romance                       10660
Comedy|Drama|Romance                 9017
Comedy|Romance                       8751
Drama|Thriller                       6513
Crime|Drama|Thriller                 5126
Crime|Drama                          4328
Horror|Thriller                      4199
Horror                               3140
Action|Adventure|Sci-Fi              2340
Action|Adventure|Sci-Fi|IMAX         2320
Drama|War                            2302
Drama|Mystery|Thriller               2275
Action|Crime|Drama|Thriller          2139
Thriller                             2132
Action|Thriller                      2070
Action|Comedy                        1943
Action|Comedy|Crime                  1841
Comedy|Crime                         1755
Action|Crime|Thriller                1619
Comedy|Crime|Drama                

__Another problem is the fact that the critics on rottentomatoes webpage probably fill in their scores for each film manually what results in a big number of score types (approx. 400 types,string format, many typing errors):__

In [60]:
list_ratings=all_critic_opinions1['CRating'].tolist()
np.unique(list_ratings)[:150]

array(["'2.25/5'", "'2.5/5'", "'2.75/5'", "'2/5'", "'3.5/5'", "'3.75/5'",
       "'3/5'", '**', '***', '****', '****1/2', '***1/2', '+',
       '+0 out of -4..+4', '+1 out of -4..+4', '+2 out of -4..+4',
       '+3 out of -4..+4', '+4 out of -4..+4', '-', '-1 out of -4..+4',
       '-1/10', '-2 out of -4..+4', '.1/5', '.2/5', '.25/4', '.5',
       '.5/ 5', '.5/10', '.5/4', '.5/5', '.65/10', '.7221/10', '.75/10',
       '.82/10', '.96/10', '0', '0 out of -4..+4', '0 out of 4..+4',
       '0 stars', '0.0/4', '0.0/4.0', '0.0/5', '0.13/1', '0.5', '0.5/10',
       '0.5/4', '0.5/4.0', '0.5/5', '0.5/5.0', '0.5/5.5', '0.52/1',
       '0.58/1', '0.59/1', '0.81/1', '0.85/10', '0/0', '0/10', '0/4',
       '0/4.0', '0/5', '0/6', '1', '1.0', '1.0/4', '1.0/4.0', '1.0/5',
       '1.0/5.0', '1.1/2', '1.2/10', '1.25/5', '1.4', '1.4/4', '1.4/4.0',
       '1.5', '1.5.4', '1.5.5', '1.5/.5', '1.5/10', '1.5/2', '1.5/3',
       '1.5/4', '1.5/4.0', '1.5/5', '1.5/5.0', '1.7/5', '1.75/4',
       '1.75/5', '1.8/

__File contains many null values for critic's scores (CRating column) that have to be dropped:__

In [64]:
all_critic_opinions1=all_critic_opinions1.dropna()

In [65]:
all_critic_opinions1.head()

CRating  TMeter         Title                Critic  Year      ID  \
0   3.5/5    0.92  Wonder Woman      alex-abad-santos  2017  108795   
2     3/4    0.92  Wonder Woman          thelma-adams  2017  108795   
3     4/5    0.92  Wonder Woman  sergio-lopez-aguirre  2017  108795   
5   3.5/4    0.92  Wonder Woman    jeffrey-m-anderson  2017  108795   
7     3/5    0.92  Wonder Woman         nigel-andrews  2017  108795   

                               Genres  
0  Action|Adventure|Animation|Fantasy  
2  Action|Adventure|Animation|Fantasy  
3  Action|Adventure|Animation|Fantasy  
5  Action|Adventure|Animation|Fantasy  
7  Action|Adventure|Animation|Fantasy

In [15]:
len(all_critic_opinions1['Critic'].unique())

1359

In [116]:
len(all_critic_opinions1['Genres'].unique())

817

__Dropping all the rows containing the "no genres" information:__

In [66]:
all_critic_opinions1=all_critic_opinions1.drop(all_critic_opinions1[all_critic_opinions1['Genres'].str.contains('no genres')].index)

__Adding additional column for grouped genres:__

In [67]:
all_critic_opinions1['Genres1']=all_critic_opinions1.Genres
all_critic_opinions1.head()

CRating  TMeter         Title                Critic  Year      ID  \
0   3.5/5    0.92  Wonder Woman      alex-abad-santos  2017  108795   
2     3/4    0.92  Wonder Woman          thelma-adams  2017  108795   
3     4/5    0.92  Wonder Woman  sergio-lopez-aguirre  2017  108795   
5   3.5/4    0.92  Wonder Woman    jeffrey-m-anderson  2017  108795   
7     3/5    0.92  Wonder Woman         nigel-andrews  2017  108795   

                               Genres                             Genres1  
0  Action|Adventure|Animation|Fantasy  Action|Adventure|Animation|Fantasy  
2  Action|Adventure|Animation|Fantasy  Action|Adventure|Animation|Fantasy  
3  Action|Adventure|Animation|Fantasy  Action|Adventure|Animation|Fantasy  
5  Action|Adventure|Animation|Fantasy  Action|Adventure|Animation|Fantasy  
7  Action|Adventure|Animation|Fantasy  Action|Adventure|Animation|Fantasy

__Genres grouping:__

In [68]:
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Children.*$)'), 'Children')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Animation.*$)'), 'Animation')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Musical.*$)'), 'Musical')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Sci-Fi.*$)'), 'Sci-Fi')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Western.*$)'), 'Western')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*War.*$)'), 'War')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Documentary.*$)'), 'Documentary')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Film-Noir.*$)'), 'Film-Noir')

In [69]:
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Thriller', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Crime|Drama|Thriller', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Horror|Thriller', 'Genres1'] = 'Horror'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Thriller', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Comedy|Drama|Romance', 'Genres1'] = 'Comedy|Romance'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Drama|Mystery|Thriller', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Drama|Thriller', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Crime|Drama', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Comedy', 'Genres1'] = 'Comedy'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Comedy|Crime', 'Genres1'] = 'Comedy'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Comedy|Crime', 'Genres1'] = 'Comedy'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Drama|Horror|Thriller', 'Genres1'] = 'Horror'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Drama|Thriller', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Crime|Thriller', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Drama', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Crime|Drama|Mystery|Thriller', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Comedy|Horror', 'Genres1'] = 'Comedy'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Crime|Drama', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Comedy|Fantasy|Romance', 'Genres1'] = 'Comedy|Romance'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Drama|Mystery', 'Genres1'] = 'Crime|Drama|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Horror|Mystery|Thriller', 'Genres1'] = 'Horror'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Drama|Fantasy|Romance', 'Genres1'] = 'Drama|Romance'

In [72]:
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Crime.*$)'), 'Crime|Thriller')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Adventure.*$)'), 'Action|Adventure')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Horror.*$)'), 'Horror')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Thriller.*$)'), 'Crime|Thriller')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*IMAX.*$)'), 'Action|Adventure')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Fantasy.*$)'), 'Fantasy')
all_critic_opinions1['Genres1'] = all_critic_opinions1.Genres1.str.replace((r'(^.*Mystery.*$)'), 'Mystery')

In [76]:
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Comedy|Romance', 'Genres1'] = 'Comedy|Romance'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Romance', 'Genres1'] = 'Comedy|Romance'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action', 'Genres1'] = 'Action|Adventure'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Comedy|Drama', 'Genres1'] = 'Comedy|Drama'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Comedy|Drama|Romance', 'Genres1'] = 'Comedy|Drama'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Drama|Romance', 'Genres1'] = 'Drama|Romance'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Romance', 'Genres1'] = 'Drama|Romance'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Mystery', 'Genres1'] = 'Crime|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Film-Noir', 'Genres1'] = 'Crime|Thriller'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Fantasy', 'Genres1'] = 'Sci-Fi'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Western', 'Genres1'] = 'Action|Adventure'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Animation', 'Genres1'] = 'Children'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Children', 'Genres1'] = 'Children|Animation'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Action|Adventure', 'Genres1'] = 'Action|Adventure|Western'
all_critic_opinions1.loc[all_critic_opinions1['Genres1']=='Sci-Fi', 'Genres1'] = 'Sci-Fi|Fantasy'

In [82]:
all_critic_opinions1.shape

(199553, 8)

__The number of film genres was lowered from 817 to 13:__

In [83]:
list_genres=all_critic_opinions1['Genres1'].tolist()
np.unique(list_genres)

array(['Action|Adventure|Western', 'Children|Animation', 'Comedy',
       'Comedy|Drama', 'Comedy|Romance', 'Crime|Thriller', 'Documentary',
       'Drama', 'Drama|Romance', 'Horror', 'Musical', 'Sci-Fi|Fantasy',
       'War'], dtype='<U24')

In [84]:
all_critic_opinions1['Genres1'].value_counts()

Crime|Thriller              38318
Drama                       26479
Sci-Fi|Fantasy              20660
Comedy                      17692
Children|Animation          15167
Comedy|Romance              14151
Action|Adventure|Western    12858
Documentary                 12080
Horror                      11550
Comedy|Drama                10428
Drama|Romance                8724
War                          6588
Musical                      4858
Name: Genres1, dtype: int64

__The number of critic scores' types was lowered to the ones presented below. The replace operation was performed manually, because due to the very low quality of data and several types of typing errors creating some regex rules would be more time consuming than simple manual replacement:__

In [126]:
all_critic_opinions1['CRating'].unique()

array([0.7 , 0.75, 0.8 , 0.87, 0.6 , 0.85, 0.5 , 0.93, 0.62, 0.9 , 1.  ,
       0.4 , 0.95, 0.82, 0.3 , 0.07, 0.37, 0.25, 0.55, 0.2 , 0.47, 0.  ,
       0.65, 0.35, 0.45, 0.1 , 0.27, 0.12, 0.48, 0.67, 0.66, 0.58, 0.88,
       0.83, 0.99, 0.57, 0.86, 0.46, 0.94, 0.44, 0.68, 0.64, 0.42, 0.63,
       0.78, 0.72, 0.76, 0.74, 0.73, 0.56, 0.43, 0.52, 0.84, 0.38, 0.53,
       0.15, 0.54, 0.17, 0.39, 0.77, 0.89, 0.34, 0.92, 0.97, 0.29, 0.98,
       0.26, 0.49, 0.96, 0.41, 0.71, 0.32, 0.28, 0.36, 0.23, 0.14, 0.13,
       0.69, 0.33, 0.24, 0.08, 0.59])

In [102]:
all_critic_opinions1["CRating"]=all_critic_opinions1.CRating.replace(['C minus',
       'Strongly Not Recommended', '+1 out of -4..+4', '82/100', '6.0/10',
       '9/20', '19/20', '10/10', '3.0/5', '85/100', 'A minus', '4.5/10',
       '6.6/10', '3.75/5', '0/4', '90/100', '58/100', '91/100', '8.8/10',
       'non-numerical', '5.0/10', '2.1/2', '88/100', '9.2/10', '8.9/10',
       'c+', '4.5/5.0', '8.3/10', '4.0/5.0', '9.85/10', '5.0/5.0',
       '9.5 out of 10', '4.0/5', '83/100', '10', '9.21/10', '2.3/4',
       '81/100', '6 out of 10', 'B plus', '8.621/10', '.5/4', '80/100'],[0.47,0.0,0.67,0.8,0.6,0.45,0.95,1.0,0.6,0.85,0.87,0.45,
                                                                        0.66,0.75,0.0,0.9,0.58,0.9,0.88,0.0,0.5,0.4,0.88,0.9,0.9,
                                                                        0.6,0.9,0.83,0.8,0.99,1.0,0.95,0.8,0.83,1.0,0.9,0.57,0.8,
                                                                        0.6,0.8,0.86,0.12,0.8])


In [117]:
all_critic_opinions1["CRating"]=all_critic_opinions1.CRating.replace(['5.78/10', '5.51/10',
       '6.95/10', '3.85/10', '20/100', '35./4', '0.0/4.0', '3/3',
       '3.20/10', '3.6/5', '8.21/10', '4.6/10', '4.7/10', '4.8/5',
       '5.5 out of 10', '***1/2', '9.78/10', '3.5 stars', '59/100',
       '7.01/10', '8.87/10', '8.6 out of 10', '5.0', '8', '4.9/10', '.5',
       '1.5/2', '8.61/10', '5.212/10', '-1 out of -4..+4', '7.0',
       '95/100', '5.70/10', '6/11', 'd-', '35/4', '36/100', '51/100',
       '9.52/10', '****', '9.121/10', '8.88/10', '+', '5.9821/10', '45',
       '9.66/10', '8.20/10', '27/100', '8.65/10', "'3.75/5'", '0.0/4',
       'high +1 out of -4..+4', '2/7', 'low +1 out of -4..+4',
       '7.3856/10', '1.5.4', '2/2', 'C +', '9.851/10', '2.0/3.0'],[0.58,0.55,0.7,0.4,0.2,0.87,0.0,1.0,0.32,0.72,0.82,0.46,0.47,
                                                                  1.0,0.55,0.7,0.98,0.7,0.6,0.7,0.88,0.86,1.0,0.8,0.5,0.5,0.75,
                                                                  0.86,0.5,0.45,0.7,0.95,0.57,0.55,0.27,0.87,0.36,0.5,0.9,0.8,
                                                                  0.9,0.89,0.6,0.6,0.45,0.97,0.82,0.27,0.86,0.8,0.0,0.67,0.28,0.67,
                                                                  0.73,0.37,1.0,0.6,0.98,0.66])

In [128]:
all_critic_opinions1.to_csv('all_critic_opinions_with_notes.csv')